### PR TITLE
Show documentation above sigs for hover

### PR DESCRIPTION
**Pull request by vigneshshankar-stripe**
With this change:
- `findDocumentation` iterates upwards using `string_view`s and accumulates documentation
- `findDocumentation` skips multi-line sig blocks, one-line sig blocks and the sig block shorthand(?) `sig do; ...; end`
- handleTextDocumentHover calls `findDocumentation` to build the final hover content 

TODO
- [x] Add separate tests for each type of response, to ensure that we retrieve the correct origin
- [x] Handle `<Class>.new` gracefully 

### Motivation

#237

### Test plan

- `lsp_test` unit tests
- `hover_*` integration tests

### CODE REVIEW

File: main/lsp/requests/completion.cc [hunk 1]
```diff
@@ -92,36 +94,94 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
...
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            do
+                it++;
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: This loop never updates `line`, so it will never find `sig do` or `end`.

Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: An alternative way to perform this parsing with only a single loop (and no inner loop) is to have state variables that the loop updates.

Example: `bool is_in_sig_block` outside of the loop. The first encounter of `end` would set it to `true`. If you encounter `end` again, we break the loop.

I believe this would also reduce the amount of code involved, since you can repurpose the existing `StartsWith(line, "sig")` conditional.

File: main/lsp/requests/completion.cc [hunk 2]
```diff
@@ -92,36 +94,94 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
...
+        }
+
+        // Handle a comment line. Do not count typing declarations.
+        else if (absl::StartsWith(line, "#") && !absl::StartsWith(line, "# typed:")) {
+            // Account for whitespace before comment e.g
+            // # abc -> "abc"
+            // #abc -> "abc"
+            int skip_after_hash = absl::StartsWith(line, "# ") ? 2 : 1;
+
+            documentation_lines.push_back(line.substr(line.find('#') + skip_after_hash, line.size()));
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: Is the `line.size()` needed? I think it'll implicitly create a string_view from position `line.find('#') + skip_after_hash` to the end of the `string_view` if you omit the second argument.

File: main/lsp/requests/hover.cc [hunk 3]
```diff
@@ -24,11 +24,38 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
     return contents;
 }
 
-unique_ptr<MarkupContent> formatRubyCode(MarkupKind markupKind, string str) {
-    if (markupKind == MarkupKind::Markdown && str.length() > 0) {
-        str = fmt::format("```ruby\n{}\n```", str);
+unique_ptr<MarkupContent> formatRubyCode(MarkupKind markupKind, string_view sigString,
+                                         optional<string_view> optionalDocString) {
+    string_view docString = optionalDocString.value_or("");
+    // Get rid of at most 1 trailing newline
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Any type of `docString` trimming should be in `findDocumentation`, not `formatRubyCode`.

Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: Also, you trim ASCII whitespace from each line in `findDocumentation`:

```
string_view line = absl::StripAsciiWhitespace(*it);
```

Does this conditional ever happen?

File: main/lsp/requests/completion.cc [hunk 4]
```diff
@@ -92,36 +94,94 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
...
+        // No other cases applied to this line, so stop looking.
+        else {
             break;
         }
-        prevLine = preDefinition.substr(prevNewlineLoc, lastNewlineLoc - prevNewlineLoc);
-        lastNewlineLoc = prevNewlineLoc;
     }
-    if (documentation.empty()) {
+
+    string documentation = absl::StrJoin(documentation_lines.rbegin(), documentation_lines.rend(), "\n");
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: I think it's safe to trim whitespace from the end of the string; WDYT? (absl has a `StripTrailingAsciiWhitespace` method.)

File: main/lsp/requests/hover.cc [hunk 5]
```diff
@@ -24,11 +24,38 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
...
+    if (docString.length() > 0 && docString.back() == '\n') {
+        docString.remove_suffix(1);
     }
-    return make_unique<MarkupContent>(markupKind, move(str));
+    if (sigString.length() > 0 && sigString.back() == '\n') {
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Ditto for `sigString`, except the trimming should be in `methodSignatureString` (and can use the `StripTrailingAsciiWhitespace` helper).

File: main/lsp/requests/hover.cc [hunk 6]
```diff
@@ -24,11 +24,38 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
...
+    string content = "";
+
+    // Add docs
+    absl::StrAppend(&content, docString);
+    if (docString.length() > 0)
+        absl::StrAppend(&content, "\n");
+
+    // Add sig
+    string formattedSigString = "";
+    if (markupKind == MarkupKind::Markdown && sigString.length() > 0) {
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Is `sigString` ever 0? Does it need to be `optional`?

File: main/lsp/requests/hover.cc [hunk 7]
```diff
@@ -24,11 +24,38 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
...
+    // Get rid of at most 1 trailing newline
+    if (docString.length() > 0 && docString.back() == '\n') {
+        docString.remove_suffix(1);
     }
-    return make_unique<MarkupContent>(markupKind, move(str));
+    if (sigString.length() > 0 && sigString.back() == '\n') {
+        sigString.remove_suffix(1);
+    }
+
+    string content = "";
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: If `findDocumentation` returns `nullopt` if the documentation string is zero-length after trimming, then we can just do something like:

```
if (optionalDocString.has_value()) {
  return absl::StrCat(optionalDocString.value(), "\n", formattedSigString);
} else {
  return formattedSigString;
}
```

File: main/lsp/requests/hover.cc [hunk 8]
```diff
@@ -24,11 +24,38 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
...
+    // Add sig
+    string formattedSigString = "";
+    if (markupKind == MarkupKind::Markdown && sigString.length() > 0) {
+        formattedSigString = fmt::format("```ruby\n{}\n```", sigString);
+    } else {
+        absl::StrAppend(&formattedSigString, sigString);
+    }
+    absl::StrAppend(&content, formattedSigString);
+
+    // Add newline at the end if needed
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Why is a newline needed?

File: main/lsp/requests/hover.cc [hunk 9]
```diff
@@ -48,20 +75,38 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
...
 
         auto resp = move(queryResponses[0]);
+
+        optional<string> documentation = nullopt;
+        if (
+                resp->isIdent() || 
+                resp->isLiteral() || 
+                resp->isConstant() || 
+                resp->isDefinition()) {
+            auto loc = resp->getTypeAndOrigins().origins[0];
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Note: `getTypeAndOrigins().origins` could potentially be zero-length, so you should guard against this possibility.

File: main/lsp/requests/hover.cc [hunk 10]
```diff
@@ -48,20 +75,38 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
         }
 
         auto resp = move(queryResponses[0]);
+
+        optional<string> documentation = nullopt;
+        if (
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: You probably don't want to `findDocumentation` for identifiers (local variables) or literals (e.g. 1, "a string", etc). Those don't generally have documentation associated with them.

File: main/lsp/requests/hover.cc [hunk 11]
```diff
@@ -48,20 +75,38 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
...
         } else if (auto defResp = resp->isDefinition()) {
             response->result = make_unique<Hover>(formatRubyCode(
-                clientHoverMarkupKind, methodDetail(*gs, defResp->symbol, nullptr, defResp->retType.type, nullptr)));
+                clientHoverMarkupKind, methodDetail(*gs, defResp->symbol, nullptr, defResp->retType.type, nullptr),
+                documentation));
         } else if (auto constResp = resp->isConstant()) {
             const auto &data = constResp->symbol.data(*gs);
             auto type = constResp->retType.type;
+            // TOOD This constResp only has a reference to where the class is referred to. Can't seem to get to the
+            // definition directly using the properties from the response.
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: I think you fixed this above by using `getTypeAndOrigin`.

File: test/helpers/position_assertions.cc [hunk 12]
```diff
@@ -818,7 +818,7 @@ bool containsLine(string_view text, string_view line) {
     }
     const bool startsOnNewLine = pos == 0 || text.at(pos - 1) == '\n';
     const bool endsLine = pos + line.size() == text.size() || text.at(pos + line.size()) == '\n';
-    return startsOnNewLine && endsLine;
+    return (startsOnNewLine && endsLine) || containsLine(text.substr(pos + line.size()), line);
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc]: Why did you need to change this method? It should be left as-is.

File: main/lsp/requests/completion.cc (hunk 1 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: Whoops -- luckily the test in lsp_test catches this.

I actually tried an approach earlier, but it required me to check that variable in other parts of the loop. It got a bit messy (e.g sig with is_in_sig_block is different from sig and ! is_in_sig_block)so I thought the inner loop makes each if block independent of what was seen before.

Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: Refreshed the line in 1d7502718, test passes now

File: main/lsp/requests/completion.cc (hunk 2 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: You're right, it's redundant. Removed in de85c6db4

File: main/lsp/requests/hover.cc (hunk 3 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: Makes sense. I think I'll ensure exactly one newline.

Now that I think about it, I don't think I should be passing in documentation into `formatRubyCode`. I'm thinking of making a separate function to concatenate sigs and docs and return MarkupContent. wdyt?

Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: For now I trimmed the documentation in newline. This actually saved quite a few lines of code in `formatRubyCode`, thank you!

2df9e0e63

File: main/lsp/requests/completion.cc (hunk 4 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: Probably won't have to because every line is stripped, but we could throw it in for good measure 👍

Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: On the topic of adding a newline for non-empty documentation, would it be advisable to use `StripTrailingAsciiWhitespace` to strip newlines? The name and [documentation](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/ascii.h#L223) don't mention it, but the abseil [tests use it](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/strip_test.cc#L178).

File: main/lsp/requests/hover.cc (hunk 5 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: 74e0d5ada

File: main/lsp/requests/hover.cc (hunk 6 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: We could make `methodSignatureString` return `optional` -- makes more sense. Could `&dispatchResult` ever be null? That's when it seems to possibly return an empty string

File: main/lsp/requests/hover.cc (hunk 7 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: This looks good, but we still need to call make_unique<MarkupContent> before return. It would look something like 
```cpp
    string content; 
    if(docString.has_value()) {
        content = absl::StrCat(docString.value(), formattedSigString);
    } else {
        content = formattedSigString;
    }
    return make_unique<MarkupContent>(markupKind, move(content));
```
With the invariant that non-empty docs always have a newline, always appending might be easier to read I think

File: main/lsp/requests/hover.cc (hunk 8 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: This was to handle docs without a newline + empty sig string. Not required anymore since we handle trimming in different functions

File: main/lsp/requests/hover.cc (hunk 9 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: !! good catch 😅 a3e1eb6c4

File: main/lsp/requests/hover.cc (hunk 11 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: Oh, yeah. 64998a434

File: test/helpers/position_assertions.cc (hunk 12 above)
Comment (vigneshshankar-stripe) [reply test/helpers/position_assertions.cc]: Ah, this was something I noticed in the hover tests. 
fc4478a8ee312dd93a9cf62d9c880c1b8854e177

If a match is found, it must match the line exactly.
Searching for `bazel` in 
```
bazelgeuse
baz
bazel
```
Would fail, although `bazel` is a valid line -- because the `pos` of `bazel` is within `bazelgeuse`.

I do think using recursion is a bit too much though. Maybe we could store all the `pos`s in a vector and use the same logic?

File: main/lsp/requests/completion.cc (hunk 4 above)
Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: If you `StrJoin` a set of empty lines, you get a string with multiple newlines -- so it can happen even though you strip whitespace from individual lines already.

Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: > would it be advisable to use StripTrailingAsciiWhitespace to strip newlines?

Yes, I expect the function to remove all whitespace. Newlines are whitespace too. :)

File: main/lsp/requests/completion.cc [hunk 13]
```diff
@@ -92,36 +94,101 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
...
+    documentation = absl::StripTrailingAsciiWhitespace(documentation);
+
+    // Ensure single newline
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: Why do you want a single newline?

Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: Also, we're guaranteed that `documentation` does not have any trailing newlines, since we stripped the trailing whitespace.

File: main/lsp/requests/hover.cc [hunk 14]
```diff
@@ -21,14 +22,28 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
         }
         start = start->secondary.get();
     }
+
+    contents = absl::StripTrailingAsciiWhitespace(contents);
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Is this needed? Does this method produce strings that have extra trailing whitespace?

File: main/lsp/requests/hover.cc (hunk 7 above)
Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: > With the invariant that non-empty docs always have a newline,

Why is this an invariant?

File: main/lsp/requests/hover.cc (hunk 6 above)
Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: I don't think `dispatchResult` is ever `null`.

File: test/helpers/position_assertions.cc (hunk 12 above)
Comment (jvilk-stripe) [reply test/helpers/position_assertions.cc]: 🤦‍♂ oh no, you're right....great catch!

Your solution works fine. Another approach is a loop; something like:

```c++
int pos = 0;
for(int pos = text.find(line, pos); pos != string::npos; pos = text.find(line, pos)) {
  const bool startsOnNewLine = pos == 0 || text.at(pos - 1) == '\n';
  const bool endsLine = pos + line.size() == text.size() || text.at(pos + line.size()) == '\n';
  if (startsOnNewLine && endsLine) {
    return true;
  }
  // Skip over newline before looping
  pos++;
}
return false;
```

File: main/lsp/requests/hover.cc (hunk 7 above)
Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: > With the invariant that non-empty docs always have a newline

I don't see the issue? `if (optionalDocString.has_value())`, then it contains a non-empty docstring. No need for a conditional to see if it's zero-length.

I think it's easier to read if we see all of the elements coming together in a single location. The fewer append operations, the better. I also think this method is the right place to append a newline after the docs, not `findDocumentation`.

File: main/lsp/requests/hover.cc [hunk 15]
```diff
@@ -48,17 +63,36 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
...
+        optional<string> documentation = nullopt;
+        if (
+                resp->isIdent() || 
+                resp->isLiteral() || 
+                resp->isConstant() || 
+                resp->isDefinition()) {
+            auto origins = resp->getTypeAndOrigins().origins;
+            if (!origins.empty()) {
+                auto loc = origins[0];
+                documentation = findDocumentation(loc.file().data(*gs).source(), loc.beginPos());
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: `loc` might not exist, so you should check `if loc.exists()` before calling `loc.file().data(*gs)`.

Sorbet uses locs with file ID 0 to indicate invalid locations internally, and calling `data()` on them will cause an exception in debug binaries and a memory error in production binaries.

File: test/helpers/position_assertions.cc (hunk 12 above)
Comment (vigneshshankar-stripe) [reply test/helpers/position_assertions.cc]: I prefer the loop actually, will probably have to do use `text.find(line.substr(pos+1))` or something

File: main/lsp/requests/hover.cc (hunk 10 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: 👍  0bcf94e2a

File: main/lsp/requests/completion.cc (hunk 4 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: :O cool, this will save me quite a few lines. f4db4d9ac

File: main/lsp/requests/hover.cc (hunk 15 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: Ah, didn't realise that. Thanks! Guarded against this and the sendResp receiver in bcb83f653

File: main/lsp/requests/hover.cc (hunk 14 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: Hmm I think this was a follow-up from https://github.com/sorbet/sorbet/pull/1499#discussion_r311180435 but now that you mention it, it doesn't seem like it would. `methodDetail`'s returns are always without newline. I'll remove this line for now -- f5207200b

File: main/lsp/requests/completion.cc (hunk 13 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: f4db4d9ac

Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: Decided to handle it in the case where documentation is not empty

File: test/helpers/position_assertions.cc (hunk 12 above)
Comment (vigneshshankar-stripe) [reply test/helpers/position_assertions.cc]: d71f1aa54

File: main/lsp/requests/completion.cc [hunk 16]
```diff
@@ -92,36 +94,100 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
...
+            if (it == all_lines.rend() || absl::StartsWith(line, "end")) {
+                break;
+            } else {
+                // Reached a sig block.
+                ENFORCE(absl::StartsWith(line, "sig do"));
+
+                // Stop looking if this is a single-line block e.g `sig do; <block>; end`
+                string_view after_sig_do = line.substr(strlen("sig do"));
+                after_sig_do = absl::StripAsciiWhitespace(after_sig_do);
+                if (absl::StartsWith(after_sig_do, ";") && absl::EndsWith(after_sig_do, "end")) {
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: I know this is faster than `absl::StartsWith("sig do;")`, but it's messy. Let's just do `absl::StartsWith(line, "sig do;") && absl::EndsWith(line, "end")`.

Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: Heck, then you can just make this:

```c++
if (it==all_ends.rend() || absl::StartsWith(line, "end") || (absl::StartsWith(line, "sig do;") && absl::EndsWith(line, "end")) { break; }
```

File: main/lsp/requests/hover.cc [hunk 17]
```diff
@@ -21,14 +22,27 @@ string methodSignatureString(const core::GlobalState &gs, const core::TypePtr &r
...
         start = start->secondary.get();
     }
+
     return contents;
 }
 
-unique_ptr<MarkupContent> formatRubyCode(MarkupKind markupKind, string str) {
-    if (markupKind == MarkupKind::Markdown && str.length() > 0) {
-        str = fmt::format("```ruby\n{}\n```", str);
+unique_ptr<MarkupContent> formatRubyCode(MarkupKind markupKind, string_view sigString,
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: Nit: `sigString` is a bit of a misnomer. It just contains a type, which might not be a `sig`. Can you rename this to `typeString`?

File: main/lsp/requests/hover.cc (hunk 6 above)
Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: You don't need to check if `sigString` is non-zero length. We should always have something; otherwise, it's a bug. You should remove this check.

File: main/lsp/requests/hover.cc (hunk 7 above)
Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: Why not:

```c++
string formattedTypeString;
/** code that sets formattedTypeString **/

return make_unique<MarkupContent>(markupKind,  absl::StrCat(docString.value_or(""), formattedTypeString));
```

That removes a bunch of LOC and `StrAppend` operations.

File: main/lsp/requests/hover.cc [hunk 18]
```diff
@@ -48,17 +62,39 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
...
+                auto loc = origins[0];
+                if (loc.exists()) {
+                    documentation = findDocumentation(loc.file().data(*gs).source(), loc.beginPos());
+                }
+            }
+        }
+
         if (auto sendResp = resp->isSend()) {
             auto retType = sendResp->dispatchResult->returnType;
+            // TODO main.method is <none> when calling .new on a class.
```
Comment (jvilk-stripe) [inline main/lsp/requests/hover.cc]: You can remove this TODO, right? We discovered in our pairing session that this is not always true, and you find docs for `new` in some scenarios.

File: main/lsp/requests/hover.cc (hunk 17 above)
Comment (jvilk-stripe) [reply main/lsp/requests/hover.cc]: Nit: `formatRubyCode` is now a misnomer. Let's rename it to `formatHoverText`.

File: test/testdata/lsp/hover.rb [hunk 19]
```diff
@@ -102,6 +129,16 @@ def main
   hash = Hash.new("default")
 # ^ hover: T::Hash[T.untyped, T.untyped]
 
+  boo = BigFoo::FOO_CONSTANT
+              # ^^^^^^^^^^^^ hover: Integer(1)
+              # ^^^^^^^^^^^^ hover: The docs for FOO_CONSTANT
+      # ^^^^^^ hover: T.class_of(BigFoo)
+      # ^^^^^^ hover: The docs for BigFoo
+  # TODO .new only works if we definte `initialize`
```
Comment (jvilk-stripe) [inline test/testdata/lsp/hover.rb]: This is OK, I think. If a class doesn't have an `initialize`, then there's no point to have docs for `new`.

Comment (jvilk-stripe) [reply test/testdata/lsp/hover.rb]: So you can remove the TODO and the nothing assertion, unless you put this here to test a regression for a bug?

File: main/lsp/requests/completion.cc (hunk 16 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/completion.cc]: Sounds good. I kept it after the iteration to keep it cleaner 6b16107c6

File: main/lsp/requests/hover.cc (hunk 17 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: Ah yes, forgot about this! thank you
90c3c22c2
91a1fd893

File: main/lsp/requests/hover.cc (hunk 18 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: b12eeff7b

File: test/testdata/lsp/hover.rb (hunk 19 above)
Comment (vigneshshankar-stripe) [reply test/testdata/lsp/hover.rb]: I'll remove the TODO since the behaviour is now expected, but I think it would be good to keep this for regression

Comment (vigneshshankar-stripe) [reply test/testdata/lsp/hover.rb]: 7fec53c50

Comment (vigneshshankar-stripe) [reply test/testdata/lsp/hover.rb]: lmk what you think -- if the behaviour is rigorous enough we could omit these 2 tests

**Comment by vigneshshankar-stripe** [comment]
Closed, made an origin branch: #1580

File: main/lsp/requests/hover.cc (hunk 7 above)
Comment (vigneshshankar-stripe) [reply main/lsp/requests/hover.cc]: 5562c3adf
